lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,288 @@
+            )
+        }
+
+        private fun validateScalarArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                actorFieldName: String
+        ): NadelSchemaValidationError? {
+            if (hydrationSourceFieldType.unwrapNonNull() is GraphQLScalarType && actorFieldArgType.unwrapNonNull() is GraphQLScalarType) {
+                if (isScalarAssignable(hydrationSourceFieldType.unwrapNonNull() as GraphQLScalarType, actorFieldArgType.unwrapNonNull() as GraphQLScalarType)) {
+                    return null
+                } else {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+            }
+            return null
+        }
+
+        private fun isScalarAssignable(typeToAssign: GraphQLNamedInputType, targetType: GraphQLNamedInputType): Boolean {
+            if (typeToAssign.name == targetType.name) {
+                return true
+            }
+            if (targetType.name == Scalars.GraphQLID.name &&
+                    (typeToAssign.name == Scalars.GraphQLString.name || typeToAssign.name == Scalars.GraphQLInt.name)) {
+                return true
+            }
+            if (targetType.name == Scalars.GraphQLString.name && typeToAssign.name == Scalars.GraphQLID.name) {
+                return true
+            }
+            return false
+        }
+
+        private fun validateListArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                hydration: UnderlyingServiceHydration,
+                actorFieldName: String
+        ): NadelSchemaValidationError? {
+            var hydrationSourceFieldInnerType: GraphQLType = hydrationSourceFieldType.unwrapNonNull().unwrapOne()
+            var actorFieldArgInnerType: GraphQLType = actorFieldArgType.unwrapNonNull().unwrapOne()
+            val errorExists = getHydrationInputErrors(
+                    hydrationSourceFieldInnerType,
+                    actorFieldArgInnerType,
+                    parent,
+                    overallField,
+                    remoteArg,
+                    hydration,
+                    actorFieldName
+            ) != null
+            if (errorExists) {
+                return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
```

Comment:
Is there any reason we don't just return the error from the call to `getHydrationInputErrors`?

This seems odd… you're getting an error, checking for its existence, then throwing it away to create a new error?

Reply:
the reason for this is because the first error is from checking the inner type,
but I wanted the error to print the type wrapped in the list in the error message, otherwise it may cause confusion for the user

Reply:
refer to this comment: https://github.com/atlassian-labs/nadel/pull/463/files/c9298a0f18682717e94b8862371a9c1cabae4cb3#r1357518727